File: recommendation-system/app/models/reaction.py
from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field


class ReactionType(str, Enum):
    LIKE = "LIKE"
    DISLIKE = "DISLIKE"


class ReactionTargetType(str, Enum):
    POST = "POST"
    COMMENT = "COMMENT"


class ReactionBase(BaseModel):
    target_type: ReactionTargetType = Field(..., alias="targetType")
    target_id: int = Field(..., alias="targetId")
    author_id: int = Field(..., alias="authorId")
    type: ReactionType

    class Config:
        populate_by_name = True


class ReactionCreate(ReactionBase):
    id: int
    created_at: datetime = Field(default_factory=datetime.now, alias="createdAt")


class ReactionResponse(ReactionBase):
    id: int
    created_at: datetime = Field(..., alias="createdAt")

    class Config:
        from_attributes = True
        populate_by_name = True

File: backend/src/main/java/com/example/Pastach/model/Reaction.java
package com.example.Pastach.model;

import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;

import java.time.Instant;

@Entity
@Table(name = "reactions", schema = "public",
        uniqueConstraints = @UniqueConstraint(name = "reactions_unique_per_target", columnNames = {"target_type", "target_id", "author_id"}),
        indexes = {
                @Index(name = "idx_reactions_target", columnList = "target_type, target_id"),
                @Index(name = "idx_reactions_author", columnList = "author_id")
        })
@Getter
@Setter
@NoArgsConstructor
public class Reaction {
    @Id
    @Setter(AccessLevel.NONE)
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    @Column(name = "target_type", nullable = false, length = 20)
    private ReactionTargetType targetType;  // POST, COMMENT

    @Column(name = "target_id", nullable = false)
    private Long targetId;  // postId, commentId

    @Column(name = "author_id", nullable = false)
    private Long authorId;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 20)
    private ReactionType type;  // LIKE, DISLIKE

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    public Reaction(ReactionTargetType targetType, Long targetId, Long authorId, ReactionType type) {
        this.targetType = targetType;
        this.targetId = targetId;
        this.authorId = authorId;
        this.type = type;
    }
}